aws/src/main/java/org/apache/iceberg/aws/glue/LockManagers.java

```diff
@@ -0,0 +1,241 @@
+    }
+
+    LockManager lockManager;
+    try {
+      lockManager = ctor.newInstance();
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(
+          String.format("Cannot initialize LockManager, %s does not implement LockManager.", impl), e);
+    }
+
+    lockManager.initialize(properties);
+    return lockManager;
+  }
+
+  abstract static class BaseLockManager implements LockManager {
+
+    private static volatile ScheduledExecutorService scheduler;
+
+    private long acquireTimeoutMs;
+    private long acquireIntervalMs;
+    private long heartbeatIntervalMs;
+    private long heartbeatTimeoutMs;
+    private int heartbeatThreads;
+
+    public long heartbeatTimeoutMs() {
+      return heartbeatTimeoutMs;
+    }
+
+    public long heartbeatIntervalMs() {
+      return heartbeatIntervalMs;
+    }
+
+    public long acquireIntervalMs() {
+      return acquireIntervalMs;
+    }
+
+    public long acquireTimeoutMs() {
+      return acquireTimeoutMs;
+    }
+
+    public int heartbeatThreads() {
+      return heartbeatThreads;
+    }
+
+    @SuppressWarnings("StaticGuardedByInstance")
+    public ScheduledExecutorService scheduler() {
+      if (scheduler == null) {
+        synchronized (this) {
+          if (scheduler == null) {
+            scheduler = Executors.newScheduledThreadPool(heartbeatThreads);
+          }
+        }
+      }
+      return scheduler;
+    }
+
+    @Override
+    public void initialize(Map<String, String> properties) {
+      this.acquireTimeoutMs = PropertyUtil.propertyAsLong(properties,
+          CatalogProperties.LOCK_ACQUIRE_TIMEOUT_MS, CatalogProperties.LOCK_ACQUIRE_TIMEOUT_MS_DEFAULT);
+      this.acquireIntervalMs = PropertyUtil.propertyAsLong(properties,
+          CatalogProperties.LOCK_ACQUIRE_INTERVAL_MS, CatalogProperties.LOCK_ACQUIRE_INTERVAL_MS_DEFAULT);
+      this.heartbeatIntervalMs = PropertyUtil.propertyAsLong(properties,
+          CatalogProperties.LOCK_HEARTBEAT_INTERVAL_MS, CatalogProperties.LOCK_HEARTBEAT_INTERVAL_MS_DEFAULT);
+      this.heartbeatTimeoutMs = PropertyUtil.propertyAsLong(properties,
+          CatalogProperties.LOCK_HEARTBEAT_TIMEOUT_MS, CatalogProperties.LOCK_HEARTBEAT_TIMEOUT_MS_DEFAULT);
+      this.heartbeatThreads = PropertyUtil.propertyAsInt(properties,
+          CatalogProperties.LOCK_HEARTBEAT_THREADS, CatalogProperties.LOCK_HEARTBEAT_THREADS_DEFAULT);
+    }
+  }
+
+  /**
+   * Implementation of {@link LockManager} that uses an in-memory concurrent map for locking.
+   * This implementation should only be used for testing,
+   * or if the caller only needs locking within the same JVM during table commits.
+   */
+  static class InMemoryLockManager extends BaseLockManager {
+
+    private static final Map<String, DefaultLockContent> LOCKS = Maps.newConcurrentMap();
+    private static final Map<String, ScheduledFuture<?>> HEARTBEATS = Maps.newHashMap();
```

Comment:
I'm fine with including this in the implementation, but it doesn't seem to me that there are that many cases where an internal heartbeat helps. If the thread that owns the lock fails, the heartbeat thread will continue to live. So the heartbeat is really a process heartbeat. And because the locks are all process-local anyway there isn't much value.

Reply:
Yes this is more for illustration purpose.